shapeshift/web
# feat: upgrade swapper protocol fee management

## Description

Upgrades the swapper architecture to support multiple protocol fees denominated in arbitrary assets, with metadata to inform the application as to what fees require balance to validate quotes. 

This change is aimed at providing future support for multi-hop trades so should show no changes from prod.

## Pull Request Type


## Issue (if applicable)

NA

## Risk

- Risk of incorrect estimated receive amounts
- Risk of incorrect protocol fees
- Risk of broken logic getting quotes

## Testing

- Check the amounts in quotes are correct - should be no changes from prod
- Check swapper logic is working correctly when getting quotes and performing trades


### Engineering

See above

### Operations

See above

## Screenshots (if applicable)

screenshot of a quote showing a protocol fee not denominated in the buy or sell assets - something that is not currently supported in production and would result in a failed transaction if there wasnt enough balance for this asset (lifi requires balance for protocol fees)
![image](https://github.com/shapeshift/web/assets/125113430/ab1017bf-b923-4802-bb71-f8220224b760)

screenshot of jumper.exchange showing the same ETH protocol fee (give or take market fluctuations)
![image](https://github.com/shapeshift/web/assets/125113430/e6071ff8-120d-4b5a-9aa1-bc4a26919721)
